src/resources/common.ts

```diff
@@ -1,41 +1,121 @@
+import {
+  Resource as ProtoResource,
+  ResourceType,
+  PolicyResource,
+  ResourceDeclareRequest,
+  ResourceDeclareResponse,
+  ActionMap,
+} from '@nitric/api/proto/resource/v1/resource_pb';
+import resourceClient from './client';
+
+export type ResourcePermissionsMap = Record<
+  string,
+  ActionMap[keyof ActionMap][]
+>;
```

Comment:
Should the structure of permissions -> actions be dictated generically?

This data structure might not suit other types of permission mapping, and implementing permsToActons as a generic method might save us some refactoring in future.

I would probably just let each of the implementations return their own action lists (even if they just use a map under the hood):
```typescript
export abstract class Resource<Permission> {
  /**
   * Unique name for the resource by type within the stack.
   *
   * This name can be used in all future references, it will be resolved automatically at runtime.
   */
  protected readonly name: string;
  public resourcePromise: Promise<ProtoResource>;

  constructor(name: string) {
    this.name = name;
  }

  setPolicies(...perms: Permission[]) {
    // Leave implementation as is (except now using this.permsToActions(...perms)).
  }

  protected abstract permsToActions(...perms: Permission[]): ActionsList;
  protected abstract register(): Promise<void>;
}
```